src/nrniv/nrnapi.h

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include "hocdec.h"
+#include "section.h"
+
+// we define these here to allow the API to be independent of internal details
+#define STACK_IS_STR 1
+#define STACK_IS_VAR 2
+#define STACK_IS_NUM 3
+#define STACK_IS_OBJVAR 4
+#define STACK_IS_OBJTMP 5
+#define STACK_IS_USERINT 6
+#define STACK_IS_SYM 7
+#define STACK_IS_OBJUNREF 8
+#define STACK_UNKNOWN -1
+
+class SectionListIterator {
+public:
+  SectionListIterator(hoc_Item *);
+  Section *next(void);
+  int done(void);
+
+private:
+  hoc_Item *initial;
+  hoc_Item *current;
+};
+
+class SymbolTableIterator {
+public:
+  SymbolTableIterator(Symlist *);
+  char const *next(void);
+  int done(void);
+
+private:
+  Symbol *current;
+};
+
+extern "C" {
+/****************************************
+ * Initialization
+ ****************************************/
+int nrn_init(int argc, const char **argv);
+void nrn_redirect_stdout(int (*myprint)(int, char *));
+
+/****************************************
+ * Sections
+ ****************************************/
+Section *nrn_new_section(char const *const name);
+void nrn_connect_sections(Section *child_sec, double child_x,
+                          Section *parent_sec, double parent_x);
+void nrn_set_section_length(Section *sec, double length);
+double nrn_get_section_length(Section *sec);
+double nrn_get_section_Ra(Section *sec);
+void nrn_set_section_Ra(Section *sec, double val);
+char const *nrn_secname(Section *sec);
+void nrn_push_section(Section *sec);
+void nrn_pop_section(void);
+void nrn_insert_mechanism(Section *sec, Symbol *mechanism);
+hoc_Item *nrn_get_allsec(void);
+hoc_Item *nrn_get_sectionlist_data(Object *obj);
+
+/****************************************
+ * Segments
+ ****************************************/
+int nrn_get_nseg(Section const *const sec);
+void nrn_set_nseg(Section *const sec, const int nseg);
+void nrn_set_segment_diam(Section *const sec, const double x,
+                          const double diam);
+double *nrn_get_rangevar_ptr(Section *const sec, Symbol *const sym,
+                             double const x);
+
+/****************************************
+ * Functions, objects, and the stack
+ ****************************************/
+Symbol *nrn_get_symbol(char const *const name);
+int nrn_get_symbol_type(Symbol *sym);
+// TODO: need a way of mapping type identifiers to meaningful names
+double *nrn_get_symbol_ptr(Symbol *sym);
+void nrn_push_double(double val);
+double nrn_pop_double(void);
+void nrn_push_double_ptr(double *addr);
+double *nrn_pop_double_ptr(void);
+void nrn_push_str(char **str);
+char **nrn_pop_str(void);
```

Comment:
Can this be `const` qualified? What are the ownership semantics of the returned value?

Reply:
Not easily (e.g. not without some kind of cast) because you'd get a:
```
error: cannot initialize return object of type 'const char **' with an rvalue of type 'char **'
```
if you tried to return a `hoc_str_pop` (which returns a `char**`) as a `char const **`, although I don't really understand why because I had thought it would be safe to add const to things, but not to remove it.

Reply:
So there are a lot of things that ought to be const but don't work without casts or modifying existing functions... for example `hoc_newobj1(Symbol*, int)`, which is used by `nrn_new_object` to create new objects doesn't actually promise that it won't change the `Symbol`.

There are also messier edge cases... The already existing `section_length` function sometimes changes the `dparam` values if a recalcultion was necessary. It's not changing the Section but it's no longer in the same state as it was.